redhat-appstudio/build-service
# Implement ability to customize build piepline selection logic

Signed-off-by: Mykola Morhun <mmorhun@redhat.com>

#### What does this PR do

This PR introduces ability to define build pipeline selection logic for components.

Build pipeline selection logic might be defined in CRs of `BuildPipelineSelector` type.
The operator analyzes CRs in the following order:
1. CR in the Application namespace that has name the same as the Application (see `Component.Spec.Application`).
2. CR named `build-pipeline-selector` in the Application namespace
3. CR named `build-pipeline-selector` in the `build-service` namespace

If no matches found or no `BuildPipelineSelector` CRs exists, then `docker-build` pipeline from default bundle is used.

#### How to test it

1. Create namespace `test`
4. Create `BuildPipelineSelector` CR named `build-pipeline-selector` in `test` namespace:
```yaml
apiVersion: appstudio.redhat.com/v1alpha1
kind: BuildPipelineSelector
metadata:
  name: build-pipeline-selector
spec:
  selectors:
    - name: Java
      pipelineRef:
        name: java-builder
        bundle: quay.io/redhat-appstudio-tekton-catalog/pipeline-java-builder:be25eed597e24e4b8404bacae3cd493b33a36226
     pipelineParams:
      - name: rebuild
        value: "true" 
     when:
        language: java
    - name: NodeJS
      pipelineRef:
        name: nodejs-builder
        bundle: quay.io/redhat-appstudio-tekton-catalog/pipeline-nodejs-builder:be25eed597e24e4b8404bacae3cd493b33a36226
      when:
        language: nodejs,node
    - name: Docker build
      pipelineRef:
        name: docker-build
        bundle: quay.io/redhat-appstudio-tekton-catalog/pipeline-docker-build:be25eed597e24e4b8404bacae3cd493b33a36226
      pipelineParams:
        - name: rebuild
          value: "true"
      when:
        dockerfile: true
    - name: Fallback
      pipelineRef:
        name: noop
        bundle: quay.io/redhat-appstudio-tekton-catalog/pipeline-docker-build:be25eed597e24e4b8404bacae3cd493b33a36226
```
5. Create application:
```yaml
apiVersion: appstudio.redhat.com/v1alpha1
kind: Application
metadata:
  name: java-application
  namespace: test
spec:
  description: Java SpringBoot basic example
  displayName: java-app-test
```
6. Fork [Java Spring Boot sample](https://github.com/devfile-samples/devfile-sample-java-springboot-basic)
7. Create `redhat-appstudio-registry-pull-secret` secret with docker image repository credentials (to push the built image into)
8. Create component (replace placeholders with your values):
```yaml
apiVersion: appstudio.redhat.com/v1alpha1
kind: Component
metadata:
  name: java-component
  namespace: test
spec:
  componentName: java-component
  application: java-application
  source:
    git:
      url: https://github.com/<username>/devfile-sample-java-springboot-basic
  containerImage: docker.io/<username>/java-component:tag
```
9. Check `build-service` operator logs and ensure that `java-builder` pipeline from the `quay.io/redhat-appstudio-tekton-catalog/pipeline-java-builder:be25eed597e24e4b8404bacae3cd493b33a36226` bundle was used. Also `rebuild` build parameter must be set to `true`.
